File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        if len(tokens) == 1:
            return int(tokens[0])

        stack = []
        r = 0
        for t in tokens:
            # print(t)
            # print(stack)
            # if t is a number,
            # push it in the stack
            if t not in ['+', '-', '*', '/']:
                stack.append(int(t))

            # else, it is an operator
            # pop 2 values and apply the operator
            else:
                b = stack.pop()
                a = stack.pop()

                # switch is 'match' in python
                # and we don't need a 'break'
                match t:
                    case '+':
                        r = a + b
                    case '-':
                        r = a - b
                    case "*":
                        r = a * b
                    case '/':
                        r = a / b
                        # The division between two integers always truncates toward zero.
                        r = math.ceil(r) if r < 0 else math.floor(r)
                
                # put this back in stack
                stack.append(r)

        # r should hold the answer,
        # also stack[-1]
        return r

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    let st = [];
    const ops = {
        '+': (a, b) => a+b,
        '-': (a,b) => a-b,
        '*': (a,b) => a*b,
        '/': (a,b) => Math.trunc(a/b)
    };
    for(let t of tokens) {
        // if t is an operand
        if(ops[t]) {
            const b = st.pop();
            const a = st.pop();
            st.push(ops[t](a,b))
        }
        // else a number
        else st.push(Number(t));
    }

    return st.pop();
};